spec.md

```diff
@@ -0,0 +1,276 @@
+
+If the directory path does not end in a `/`, a `/` is appended and the user is redirected. This helps avoid serving duplicate content from different paths.
+
+Etags and range requests are supported. File mimetypes are detected and set in `Content-Type`.
+
+The header `X-IPFS-Path` is set to the URL path. `IPFS-Hash` is set to CID of the IPFS path provided, this may differ from the CID in the URL path.
+
+### HEAD `/ipfs/<CID>[/<path>]`
+
+The same as GET, but with no response body.
+
+### POST `/ipfs/`
+
+Upload file(s) to IPFS.
+
+#### Headers
+- `X-IPFS-Pin` pin the uploaded file(s) if this header is present
+
+#### Response
+The response body is uploaded, and a 201 Created redirect to `ipfs://<CID>` is returned. The response header `IPFS-Hash` is set to the CID of the uploaded content.
+
+Form data is supported, in which case multiple files could be uploaded, and the returned CID would point to a directory containing those files.
+
+### PUT `/ipfs/<CID>[/<path>]`
+
+Upload a file to an existing path, such as `/ipfs/<CID>/path/to/file.ext`. Form data is supported, in which case the path must point to a directory, such as `/ipfs/<CID>/path/to/dir`.
+
+This can also be used to wrap a file in a directory, by using the empty directory CID: `/ipfs/bafybeiczsscdsbs7ffqz55asqdf3smv6klcw3gofszvwlyarci47bgf354`.
+
+#### Headers
+- `X-IPFS-Pin` pin the resulting new CID if this header is present
+
+#### Response
+A 201 Created redirect to `ipfs://<new root CID>[/<path>]` is returned. The response header `IPFS-Hash` is set to the new root CID that has resulted from the
+file(s) upload.
+
+### DELETE `/ipfs/<CID>/<path>`
+
+Remove the file or directory at the provided path.
+
+#### Headers
+- `X-IPFS-Pin` pin the new CID that results from deletion if this header is present
+
+#### Response
+A 201 Created redirect to `ipfs://<new root CID>[/<path>]` is returned. The path in the redirect is the directory containing the deletion path. So if the path in the URL was `/ipfs/<CID>/path/to/file.ext`, the redirect will be to `ipfs://<new CID>/path/to`.
+
+The response header `IPFS-Hash` is set to the new root CID that has resulted from the deletion.
+
+### DELETE `/ipfs/<CID>`
+
+Unpin the provided CID.
+
+#### Response
+200 OK if successful.
+
+
+## IPNS APIs
+
+
+### GET or HEAD `/ipns/<key/domain>[/<path>]`
+
+#### Options
+Same as GET `/ipfs/<CID>[/<path>]`
+
+#### Response
+Resolves the IPNS key to an IPFS path, and then returns the same response as `/ipfs/`. Also supports resolving custom domains with [DNSLink](https://dnslink.io/).
+
+The header `X-IPFS-Path` is set to the URL path. `IPFS-Hash` is set to CID of the IPFS path provided, this may differ from the CID in the URL path. This allows for getting the CID of an IPNS path/key.
```

Comment:
```suggestion
The header `X-IPFS-Path` is set to the URL path. `X-IPFS-Hash` is set to CID of the IPFS path provided, this may differ from the CID in the URL path. This allows for getting the CID of an IPNS path/key.
```

Reply:
It's `IPFS-Hash`, not `X-IPFS-Hash`. This is a header that was in the codebase before I started adding my own things, so I think it shouldn't be changed.

Reply:
Should we be using `IPFS-Path` instead of `X-IPFS-Path`? Or was that also a preexisting header?

Reply:
That was also preexisting.